source/common/stats/symbol_table_impl.h

```diff
@@ -0,0 +1,122 @@
+#pragma once
+
+#include <algorithm>
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "envoy/common/exception.h"
+#include "envoy/stats/symbol_table.h"
+
+#include "common/common/assert.h"
+
+#include "absl/strings/str_join.h"
+#include "absl/strings/str_split.h"
+
+namespace Envoy {
+namespace Stats {
+
+/**
+ * Underlying SymbolTableImpl implementation which assists with per-symbol reference counting.
+ *
+ * The underlying Symbol / SymbolVec data structures are not meant for public consumption. One side
+ * effect of the non-monotonically-increasing symbol counter is that if a string is encoded, the
+ * resulting stat is destroyed, and then that same string is re-encoded, it may or may not encode to
+ * the same underlying symbol.
+ */
+class SymbolTableImpl : public SymbolTable {
+public:
+  StatNamePtr encode(const std::string& name) override;
+
+  // For testing purposes only.
+  size_t size() const override {
+    RELEASE_ASSERT(encode_map_.size() == decode_map_.size(), "");
+    return encode_map_.size();
+  }
+
+private:
+  friend class StatNameImpl;
+  friend class StatNameTest;
+  /**
+   * Decodes a vector of symbols back into its period-delimited stat name.
+   * If decoding fails on any part of the symbol_vec, we release_assert and crash hard, since this
+   * should never happen, and we don't want to continue running with a corrupt stats set.
+   *
+   * @param symbol_vec the vector of symbols to decode.
+   * @return std::string the retrieved stat name.
+   */
+  std::string decode(const SymbolVec& symbol_vec) const;
+
+  /**
+   * Since SymbolTableImpl does manual reference counting, a client of SymbolTable (such as
+   * StatName) must manually call free(symbol_vec) when it is freeing the stat it represents. This
+   * way, the symbol table will grow and shrink dynamically, instead of being write-only.
+   *
+   * @param SymbolVec& the vector of symbols to be freed.
+   * @return bool whether or not the total free operation was successful. Expected to be true.
+   */
+  void free(const SymbolVec& symbol_vec);
+
+  /**
+   * Convenience function for encode(), symbolizing one string segment at a time.
+   *
+   * @param std::string& the individual string to be encoded as a symbol.
```

Comment:
The Envoy convention (which has an unknown relationship to actual doxygen :) is you need here the name of the parameter, rather than the type of it.  In contrast, the return annotation needs the type, and there is no name...